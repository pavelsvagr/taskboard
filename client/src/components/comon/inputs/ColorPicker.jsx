import React, { useState } from "react"

import { Button, Col, Modal, Row } from "antd"
import PropTypes from "prop-types"

const ColorPicker = ({ value, onChange, data }) => {
  const [showModal, setShowModal] = useState()

  const handleChange = (color) => {
    setShowModal(false)
    if (onChange) {
      onChange(color)
    }
  }

  return (
    <div>
      {showModal && (
        <Modal
          width={400}
          visible
          title="Select color"
          onCancel={() => setShowModal(false)}
          footer={false}
        >
          <Row gutter={[24, 24]}>
            {data.map((color) => (
              <Col key={color} span={4} style={{ textAlign: "center" }}>
                {value === color ? (
                  <Button
                    className={`color-picker--selected color-${color}`}
                    shape="circle"
                    size="large"
                    icon={<span />}
                    ghost
                    disabled
                  />
                ) : (
                  <Button
                    className={`color-picker color-${color}`}
                    size="large"
                    shape="circle"
                    ghost
                    onClick={() => handleChange(color)}
                    icon={<span />}
                  />
                )}
              </Col>
            ))}
          </Row>
        </Modal>
      )}
      <div>
        <Button
          className={`fill color-${value}`}
          onClick={() => setShowModal(true)}
          icon={<span />}
        />
      </div>
    </div>
  )
}

ColorPicker.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func,
  data: PropTypes.arrayOf(PropTypes.string),
}

ColorPicker.defaultProps = {
  value: null,
  data: [],
  onChange: null,
}

export default ColorPicker
