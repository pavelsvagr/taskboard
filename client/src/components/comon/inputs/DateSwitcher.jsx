import React, { Component } from "react"
import moment from "moment"
import momentTypes from "react-moment-proptypes"

import { Button, Col, DatePicker, Row } from "antd"
import { LeftOutlined, RightOutlined } from "@ant-design/icons"

import renderBoardDate from "@shared/format/renderBoardDate"
import PropTypes from "prop-types"

class DateSwitcher extends Component {
  getDatePickerType = (intervals) => {
    switch (intervals) {
      case "weeks":
        return "week"
      case "months":
        return "month"
      default:
        return "date"
    }
  }

  getFormat = (intervals) => {
    switch (intervals) {
      case "weeks":
        return (value) => renderBoardDate(value, intervals)
      case "months":
        return "MM. YYYY"
      default:
        return "DD. MM. YYYY"
    }
  }

  handleNext = () => {
    const { selected, onChange, intervals } = this.props
    const actual = moment(selected)
    const next = actual.add(1, intervals)
    onChange(next)
  }

  handlePrevious = () => {
    const { selected, onChange, intervals } = this.props
    const actual = moment(selected)
    const prev = actual.subtract(1, intervals)
    onChange(prev)
  }

  render() {
    const {
      intervals,
      selected,
      onChange,
      disabled,
      buttonProps,
      ...otherProps
    } = this.props

    return (
      <div>
        <Row gutter={[4, 0]} style={{ justifyContent: "center" }}>
          <Col>
            <Button
              onClick={this.handlePrevious}
              icon={<LeftOutlined />}
              disabled={disabled}
              {...buttonProps}
            />
          </Col>
          <Col>
            <DatePicker
              {...otherProps}
              disabled={disabled}
              value={selected}
              isReadOnly
              allowClear={false}
              format={this.getFormat(intervals)}
              picker={this.getDatePickerType(intervals)}
              onChange={onChange}
            />
          </Col>
          <Col>
            <Button
              onClick={this.handleNext}
              icon={<RightOutlined />}
              disabled={disabled}
              {...buttonProps}
            />
          </Col>
        </Row>
      </div>
    )
  }
}

DateSwitcher.propTypes = {
  selected: PropTypes.oneOfType([momentTypes.momentObj, PropTypes.string]),
  onChange: PropTypes.func,
  intervals: PropTypes.string,
  disabled: PropTypes.bool,
  buttonProps: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.string])
  ),
}

DateSwitcher.defaultProps = {
  selected: null,
  onChange: null,
  disabled: false,
  buttonProps: {},
  intervals: null,
}

export default DateSwitcher
