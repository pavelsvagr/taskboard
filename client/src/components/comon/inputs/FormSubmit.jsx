import React from "react"
import { Button } from "antd"
import PropTypes from "prop-types"

function FormSubmit({ title, size }) {
  return (
    <div className="form__submit">
      <Button htmlType="submit" type="primary" size={size}>
        {title}
      </Button>
    </div>
  )
}

FormSubmit.propTypes = {
  title: PropTypes.string.isRequired,
  size: PropTypes.string,
}

FormSubmit.defaultProps = {
  size: "medium",
}

export default FormSubmit
