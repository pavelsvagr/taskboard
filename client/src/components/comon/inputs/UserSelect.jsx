import React, { Component } from "react"
import { Select, Spin } from "antd"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import debounce from "lodash.debounce"
import { fetchUsers, SUBJECT_USERS } from "../../../actions"
import shapes from "../../../types"
import DataPagination from "../../board/forms/DataPagination"
import EmptyData from "../data/EmptyData"

class UserSelect extends Component {
  constructor(props) {
    super(props)

    this.handleFetch = debounce(props.fetchUsers, 600)

    this.state = {
      search: "",
    }
  }

  handleSearch = (search) => {
    const { users } = this.props
    const { limit, offset } = users
    this.setState({ search })
    this.handleFetch(search, offset, limit)
  }

  render() {
    const {
      value,
      onChange,
      users,
      disabled,
      loading,
      fetchUsers: fetchAction,
    } = this.props
    const { limit = 0, count = 0, offset = 0, data = [] } = users
    const { search } = this.state

    return (
      <Select
        mode="multiple"
        onChange={onChange}
        onSearch={this.handleSearch}
        searchValue={search}
        value={value}
        placeholder="Add users"
        optionLabelProp="label"
        optionFilterProp="search"
        dropdownRender={(menu) => (
          <Spin spinning={!!loading?.states[SUBJECT_USERS]}>
            {data.length ? (
              <>
                {menu}
                {limit < count && (
                  <div className="text-center p-md">
                    <DataPagination
                      limit={limit}
                      offset={offset}
                      count={count}
                      search={search}
                      onFetch={fetchAction}
                    />
                  </div>
                )}
              </>
            ) : (
              <EmptyData title="No user found" />
            )}
          </Spin>
        )}
      >
        {data.map((user) => (
          <Select.Option
            key={user._id}
            value={user._id}
            disabled={disabled[user._id]}
            label={user.name}
            search={`${user.email}${user.name}`}
          >
            {user.email} 
            {' '}
            {user.name}
          </Select.Option>
        ))}
      </Select>
    )
  }
}

UserSelect.propTypes = {
  value: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func,
  users: shapes.paginate(shapes.user),
  disabled: PropTypes.objectOf(PropTypes.bool),
  fetchUsers: PropTypes.func.isRequired,
  loading: shapes.loading,
}

UserSelect.defaultProps = {
  value: null,
  onChange: null,
  users: {},
  disabled: [],
  loading: null,
}

function mapStateToProps({ users, loading }) {
  return { users: users.all, loading }
}

export default connect(mapStateToProps, { fetchUsers })(UserSelect)
