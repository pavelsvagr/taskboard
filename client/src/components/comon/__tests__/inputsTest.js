/* eslint react/jsx-filename-extension:0 */
import React from "react"
import renderer from "react-test-renderer"
import moment from "moment"
import {Provider} from "react-redux"
import configureMockStore from "redux-mock-store"
import thunk from "redux-thunk"
import windowMedia from "../../../helpers/tests/windowMedia"
import ColorPicker from "../inputs/ColorPicker"
import DateSwitcher from "../inputs/DateSwitcher"
import FormSubmit from "../inputs/FormSubmit"
import UserSelect from "../inputs/UserSelect"
import testStore from "../../../helpers/tests/testAdminStore"

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe("Inputs", () => {
  it("Snapshot: ColorPicker.jsx", () => {
    Object.defineProperty(window, "matchMedia", windowMedia)
    const tree = renderer
      .create(<ColorPicker data={["white", "blue"]} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("Snapshot: DateSwitcher.jsx", () => {
    Object.defineProperty(window, "matchMedia", windowMedia)
    const tree = renderer
      .create(
        <DateSwitcher
          intervals="days"
          selected={moment("2020-10-01")}
          onChange={jest.fn()}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("Snapshot: FormSubmit.jsx", () => {
    Object.defineProperty(window, "matchMedia", windowMedia)
    const tree = renderer.create(<FormSubmit title="test submit" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("Snapshot: UserSelect.jsx", () => {
    const store = mockStore(testStore)
    Object.defineProperty(window, "matchMedia", windowMedia)
    const tree = renderer
      .create(
        <Provider store={store}>
          <UserSelect
            value={[]}
            onChange={jest.fn()}
          />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
