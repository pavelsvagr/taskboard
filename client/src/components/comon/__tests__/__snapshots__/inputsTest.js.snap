// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputs Snapshot: ColorPicker.jsx 1`] = `
<div>
  <div>
    <button
      className="ant-btn ant-btn-icon-only fill color-null"
      onClick={[Function]}
      type="button"
    >
      <span />
    </button>
  </div>
</div>
`;

exports[`Inputs Snapshot: DateSwitcher.jsx 1`] = `
<div>
  <div
    className="ant-row"
    style={
      Object {
        "justifyContent": "center",
        "marginLeft": -2,
        "marginRight": -2,
      }
    }
  >
    <div
      className="ant-col"
      style={
        Object {
          "paddingLeft": 2,
          "paddingRight": 2,
        }
      }
    >
      <button
        className="ant-btn ant-btn-icon-only"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label="left"
          className="anticon anticon-left"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="left"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      className="ant-col"
      style={
        Object {
          "paddingLeft": 2,
          "paddingRight": 2,
        }
      }
    >
      <div
        className="ant-picker"
        onMouseUp={[Function]}
      >
        <div
          className="ant-picker-input"
        >
          <input
            autoComplete="off"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            placeholder="Select date"
            readOnly={true}
            size={14}
            title=""
            value=""
          />
          <span
            className="ant-picker-suffix"
          >
            <span
              aria-label="calendar"
              className="anticon anticon-calendar"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="calendar"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      className="ant-col"
      style={
        Object {
          "paddingLeft": 2,
          "paddingRight": 2,
        }
      }
    >
      <button
        className="ant-btn ant-btn-icon-only"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label="right"
          className="anticon anticon-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Inputs Snapshot: FormSubmit.jsx 1`] = `
<div
  className="ant-row ant-form-item form__submit"
  style={Object {}}
>
  <div
    className="ant-col ant-form-item-control"
    style={Object {}}
  >
    <div
      className="ant-form-item-control-input"
    >
      <div
        className="ant-form-item-control-input-content"
      >
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="submit"
        >
          <span>
            test submit
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Inputs Snapshot: UserSelect.jsx 1`] = `
<div
  className="ant-select ant-select-multiple ant-select-show-search"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
>
  <div
    className="ant-select-selector"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <span
      className="ant-select-selection-search"
      style={
        Object {
          "width": 0,
        }
      }
    >
      <input
        aria-activedescendant="undefined_list_0"
        aria-autocomplete="list"
        aria-controls="undefined_list"
        aria-haspopup="listbox"
        aria-owns="undefined_list"
        autoComplete="off"
        className="ant-select-selection-search-input"
        onChange={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onPaste={[Function]}
        readOnly={true}
        role="combobox"
        style={
          Object {
            "opacity": 0,
          }
        }
        type="search"
        unselectable="on"
        value=""
      />
      <span
        aria-hidden={true}
        className="ant-select-selection-search-mirror"
      >
        
        Â 
      </span>
    </span>
    <span
      className="ant-select-selection-placeholder"
    >
      Add users
    </span>
  </div>
</div>
`;
